### ISSUE TITLE
DateTime.Kind does not trigger entity changed

### ISSUE THREAD

**Issue by ddrenkovski** [dotnet/efcore#31240]
Is it done "by design" for a DateTime Property on an Entity not to be seen as modified when its Kind changes?

The issue was already [asked ](https://github.com/dotnet/runtime/issues/88729#issue-1800805991) to the runtime team. But would like to know if EFCore team has considered this behavior and if there is a workaround?

**Comment by roji** [comment]
Duplicate of #88729

**Comment by roji** [comment]
@ddrenkovski note that in general, EF does not persist the Kind property into the database; for example, [SQL Server's datetime2 type](https://learn.microsoft.com/en-us/sql/t-sql/data-types/datetime2-transact-sql?view=sql-server-ver16) is simply a date and time in the database, without any data indicating whether it's UTC or otherwise (this is the same with most other database timestamp types AFAIK). So even if EF were modified to take the Kind into account for change tracking purposes, you still wouldn't be able to actually persist that information.

**Comment by ajcvickers** [comment]
@ddrenkovski You can implement [a custom value comparer](https://learn.microsoft.com/en-us/ef/core/modeling/value-comparers?tabs=ef5) that will cause EF to detect this, although as Shay says, there is little point.

**Comment by ddrenkovski** [comment]
Thank you @ajcvickers that is exactly what I needed (for my use case)

**Comment by roji** [comment]
Out of curiosity, can you please explain what this is useful for, given that the Kind isn't persisted to the database, as explained above?

**Comment by ddrenkovski** [comment]
@roji  I have a Postgesql database and it complained when trying to save a DateTime value without Kind set to UTC. So to correct that error I added database wide conversion for the DateTime type

       protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
        {
            // Configure ALL the DateTime properties in our EF model so they'll have DateTime.Kind set to Utc. 
            configurationBuilder
                .Properties<DateTime>()
                .HaveConversion<DateTimeToUtcConverter, DateTimeComparer>();
        }

where:

    /// <summary>
    /// Deserializes DateTime so if its Kind is Unspecified then it's set to Utc. 
    /// </summary>
    public class DateTimeToUtcConverter : ValueConverter<DateTime, DateTime>
    {
        public DateTimeToUtcConverter() : base(v => DateTime.SpecifyKind(v, DateTimeKind.Utc), v => DateTime.SpecifyKind(v, DateTimeKind.Utc))
        {
        }
    }
    
    /// <summary>
    /// Overwrites the default DateTime compare. 
    /// </summary>
    public class DateTimeComparer : ValueComparer<DateTime>
    {
        public DateTimeComparer() : base((d1,d2) => d1 == d2 && d1.Kind == d2.Kind, d => d.GetHashCode(), d => d)
        {
        }
    }

So now every time it is read or written to the database it returns a DateTime with Kind UTC, and if during **comparison of values**  the kind is different, it forces EF to consider it as a change.

NOTE: comparison of values is used when parsed json values arrive and we compare them with existing database values, and instead having to add UTC to any parsed value, this is a better workaround.

Not sure if the whole logic is understandable with this, but hopefully you get the gist. Thank you for investing time into this.

**Comment by roji** [comment]
@ddrenkovski I'm a bit confused by the above...

The use of a value **converter** to ensure that the DateTime Kind is UTC makes sense, since PostgreSQL requires that when inserting values into columns of type `timestamp with time zone` (the default for DateTime). That's all good and well.

But the use of a value **comparer** here is unclear to me. Comparers are used when an entity has been loaded from the database, in order to know if a property was changed so that the change can be persisted back to the database when SaveChanges is called. However, in this case, DateTimes loaded from PG will always already have Kind UTC - so why would you need a comparer to check if the Kind changed?

**Comment by ddrenkovski** [comment]
@roji I suspected that the explanation might be too little for the second part :)
There are two hash methods that calculate before and after object:

1. There is a logic to calculate hash of an object **Foo** that might be in database or not. When there is data from the database then a hash value is calculated based on some properties. Now we check if there is a need to update the data of that object in database from an incoming json. 

2. The json is deserialized to **Foo** and there is some logic where id and timestamps are remove before the creation of the hash. 
The json also contains properties with valid date properties, but it doesn't have a timezone attach to it (but the third party claims that all dates are utc). There is a way to convert each datetime property to have UTC Kind that requires on each property where there is a date involved to do DateTime.SpecifyKind(foo.property, DateTimeKind.Utc). But this may lead to repetition of DateTime.SpecifyKind and in future if another date property is added the person has to know why it is needed to do the conversion (they'll get an error, only when testing). To avoid this, the solution which I came up with is:

       private void ReadDateTimeAsUTC(EntityBase<long>? entity)
        {
            if (entity == null) { return; }

            context.Entry(entity).State = EntityState.Detached;
            foreach (var prop in context.Entry(entity).Properties)
            {
                if (prop.Metadata.ClrType != typeof(DateTime)) { continue; }
                // setting the kind works because we have overwritten the DateTime comparison logic called by EF
                prop.CurrentValue = DateTime.SpecifyKind((DateTime)prop.CurrentValue!, DateTimeKind.Utc);
            }
        }
This way we only write once DateTime.SpecifyKind for all Properties of the entity. If the dates are the same, but from the deserialized **Foo** we have no kind, but from the database readout we have Utc kind, then the above code not work, because EF will see them also as same without the need to modify the CurrentValue. And because of that the calculated hash will differ. With the above logic, when the **Foo** object is created from the json and then hashed, it gives the same hash value as the hash value from data read from the database.


Hopefully it is a bit clearer. Do you maybe have a different suggestion/solution/approach?

**Comment by roji** [comment]
So first, the root problem here seems to be that when you serialize your incoming JSON documents to Foo, you're getting DateTimes with Kind Unspecified rather than UTC. If that's the case, I'd fix that specific issue by using a [System.Text.Json JsonConverter](https://learn.microsoft.com/en-us/dotnet/standard/datetime/system-text-json-support#custom-support-for--and-). In other words, fix your problem early by cleaning up/normalizing your data when it enters your application, rather than letting it leak into your application (e.g. EF setup). Note that this is done only once for all DateTimes being deserialized (the DateTime.SpecifyKind), so you don't need to keep track of new properties or anything.

If you'd rather do the conversion in EF for some reason (less recommended!), you can still do that by setting up a value converter for *all* DateTime properties in your model - [see this doc page](https://learn.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration). This again puts DateTime.SpecifyKind in a single, global place and doesn't need to be done per-property.

Other than that, I'm still not sure at which point ReadDateTimeAsUTC is called and what exactly it's supposed to do, or how a value comparer fits into all this... But do one of the above suggestions make sense for you?

**Comment by ddrenkovski** [comment]
Thank you for the suggestions. 
1. I previously gave it a try with JsonConverter, but it didn't work. Don't remember the exact reason, but maybe I should try again since (I agree) it is a more reasonable approach.

2. Now I didn't understand you, what you meant with the second paragraph. I already have ` .HaveConversion<DateTimeToUtcConverter, DateTimeComparer>();` or are you suggesting something different?

3. The ReadDateTimeAsUTC method is called before the hash function is called. Let me try to explain that part a bit more:

The simplified object **Foo** that is hashed has some properties:

   public class Foo
   {
         long Id;
         string Name;
         DateTime OrderDate;
         DateTime DateCreated;
         DateTime DateModified;
   }

Now before I hash it, I set the id= 0, DateCreated =DateTime.MinValue, and  DateModified =DateTime.MinValue (remark: this are part of an **BaseEntity**). The other properties (Name and OrderDate) are needed for comparison reason - for the hash. After the "unneeded" properties are defaulted, a call is made to ReadDateTimeAsUTC  to "fix" the DateTime properties (OrderDate in this example). I use the same type of object as it is used for the database (Entity), to deserialize the json to it (maybe this is also a mistake, and I should use dto). Because the object to which the json is deserialized to is also an entity, that is why the value comparer is triggered.

**Comment by roji** [comment]
> The ReadDateTimeAsUTC method is called before the hash function is called. Let me try to explain that part a bit more:

OK, but I still don't understand what this has to do with persisting any change back with SaveChanges or why a value comparer is needed. Are you calling ReadDateTimeAsUTC after deserializing something from JSON which isn't yet tracked, or on a tracked entity instance just loaded from EF? If it's deserialized from JSON, and you're attaching the JSON-deseriailzed instance, then the entity should be attached (so I'm not sure why the state is set to Detached in your code above).

In any case, at the end of the day, I simply don't see why the Kind of the DateTime should matter for EF value comparison, given that the database column doesn't hold it... If the DateTime is the same except for the Kind, why would you want to persist it to the database, given that you'll persist exactly the same thing that's already in the database?

**Comment by ddrenkovski** [comment]
> on a tracked entity instance just loaded from EF

Yes. First if there is an entity it is loaded from DB, it is cleaned up (`"unneeded" properties are defaulted`), DateTime "fixed", and hash is calculated. Then on the same entity instance that is fetched from db the parsed data from the json is added (mapped on to it). During this **mapping of the values**, the EF does its own comparison if the fields are changed (regardless if state is Detached (as I tried it)). By default the DateTime fields are considered the same, even though the Kind is different. But if you serialize them to json and get the  hash code then it is different ( see [fiddle ](https://dotnetfiddle.net/uHwMMu) ).

(per your second paragraph)
 I am not that familiar with Postresql

> I simply don't see why the Kind of the DateTime should matter for EF value comparison, given that the database column doesn't hold it.. 

but you also mentioned:
> since PostgreSQL requires that when inserting values into columns of type timestamp with time zone (the default for DateTime).

that is why the Kind type issues occur for me in the first place

Sorry if I didn't manage to convey the logic clearly enough.

You are probably right, I may have done it a bit too complicated, and need to refactor it. For the moment it works as intended.

**Comment by roji** [comment]
It does sound like you arrived at a pretty complicated setup there... I against recommend trying to get UTC DateTimes from your JSON deserialization - this would ensure you only have one Kind everywhere in your application (UTC), and you can forget about anything else. Otherwise, you can use EF converters to e.g. do DateTime.SpecifyKind() to set Kind=UTC when writing to the database, but also set Kind=Unspecified when reading from the database. This would ensure that in .NET you only have Unspecified DateTimes.

>I am not that familiar with Postresql
>
>>I simply don't see why the Kind of the DateTime should matter for EF value comparison, given that the database column doesn't hold it..
>
>but you also mentioned:
>
> since PostgreSQL requires that when inserting values into columns of type timestamp with time zone (the default for DateTime).

There's some confusion here... Npgsql requires that all values inserted into `timestamp with time zone` columns have Kind=Utc with writing DateTimes to the database; it also returns DateTime instances with Kind=Utc when reading from the database. The database column **does not** hold any Kind information - the PG column **type** (`timestamp with time zone`) is the thing that declares that its contents are UTC. So again, the Kind is **never** stored in the database.

Because of this, you should never need a value comparer which inspects the DateTime.Kind.
